kernel/memory/src/paging/mapper.rs

```diff
@@ -867,3 +867,104 @@ pub fn mapped_pages_unmap(
+        // ✅ The lifetime of the returned reference `&T` is tied to the lifetime of the `MappedPages`,
+        //    ensuring that the `MappedPages` object will persist at least as long as the reference.
+        unsafe { self.ptr.as_ref() }
+    }
+}
+impl<T: FromBytes> BorrowedMappedPages<T> {
+    /// Immutably borrows the given `MappedPages` as an instance of type `&T` 
+    /// starting at the given `offset` into the `MappedPages`.
+    ///
+    /// See [`MappedPages::as_type()`] for more info.
+    ///
+    /// Returns an error containing the unmodified `MappedPages` and a string
+    /// describing the error.
+    pub fn try_into_borrowed(
+        mp: MappedPages,
+        offset: usize,
+    ) -> Result<BorrowedMappedPages<T>, (MappedPages, &'static str)> {
+        let mut borrowed_mp = BorrowedMappedPages {
+            ptr: NonNull::<T>::dangling(),
+            mp,
+        };
+        borrowed_mp.ptr = match borrowed_mp.mp.as_type::<T>(offset) {
+            Ok(r) => r.into(),
+            Err(e_str) => return Err((borrowed_mp.mp, e_str)),
+        };
+        Ok(borrowed_mp)
+    }
+
+    /// Consumes this `BorrowedMappedPages` and returns the inner `MappedPages`.
+    pub fn into_inner(self) -> MappedPages {
+        self.mp
+    }
+}
+
+/// A mutably borrowed `MappedPages` object that derefs to `&T` and `&mut T`.
+///
+/// When dropped, the borrow ends and the contained `MappedPages` is dropped and unmapped.
+pub struct BorrowedMutMappedPages<T: FromBytes> {
+    ptr: NonNull<T>,
+    mp: MappedPages,
+}
+impl<T: FromBytes> Deref for BorrowedMutMappedPages<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        // SAFETY:
+        // ✅ Same as `BorrowedMappedPages<T>`.
+        unsafe { self.ptr.as_ref() }
+    }
+}
+impl<T: FromBytes> core::ops::DerefMut for BorrowedMutMappedPages<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        // SAFETY:
+        // ✅ Same as `BorrowedMappedPages<T>`, plus:
+        // ✅ The underlying `MappedPages` is guaranteed to be writable by `MappedPages::as_type_mut()`.
+        unsafe { self.ptr.as_mut() }
+    }
+}
+impl<T: FromBytes> BorrowedMutMappedPages<T> {
+    /// Mutably borrows the given `MappedPages` as an instance of type `&mut T` 
+    /// starting at the given `offset` into the `MappedPages`.
+    /// 
+    /// See [`MappedPages::as_type_mut()`] for more info.
+    /// 
+    /// Returns an error containing the unmodified `MappedPages` and a string
+    /// describing the error.
+    pub fn try_into_borrowed_mut(
+        mp: MappedPages,
+        offset: usize,
+    ) -> Result<BorrowedMutMappedPages<T>, (MappedPages, &'static str)> {
+        let mut borrowed_mp = BorrowedMutMappedPages {
+            ptr: NonNull::<T>::dangling(),
+            mp,
+        };
+        borrowed_mp.ptr = match borrowed_mp.mp.as_type_mut::<T>(offset) {
+            Ok(r) => r.into(),
+            Err(e_str) => return Err((borrowed_mp.mp, e_str)),
+        };
```

Comment:
```suggestion
        let mut borrowed_mp = BorrowedMutMappedPages {
            ptr: match borrowed_mp.mp.as_type_mut::<T>(offset) {
                Ok(r) => r.into(),
                Err(e_str) => return Err((borrowed_mp.mp, e_str)),
            },
            mp,
        };
```
Same as above.

Reply:
But the `borrowed_mp` variable doesn't exist yet in your snippet at line 956, does it?

Reply:
Same as above.